### ISSUE TITLE
Automated script can break other kernels as well

### ISSUE THREAD

**Issue by LecrisUT** [Frogging-Family/linux-tkg#303]
I'm not sure which part of the script is at fault, but it manged to make all of the installed kernels become bricked. Network, bluetooth, most functionalities do not work anymore. So trying to use `./install.sh uninstall-helper` naturally doesn't do anything.

Any ideas of what might cause this? I was just trying to add `anbox` patches, but it might be due to the myriad of other options presented. Still why would it affect other kernels?

**Comment by AdelKS** [comment]
Hello,

`install.sh` should not touch any other file from other kernels and the `uninstall-help` command only gives you want you need to delete to remove a kernel. Can you give the output of that command nonetheless ? What distro are you running on ? From the looks of it, your other kernels lost their modules folder in `/lib/modules` so if you boot one of them, no module is found and thus you cannot use anything that relies on them.

**Comment by LecrisUT** [comment]
I would have to back-up and retry the steps. If you have some good ways to do this backup I'll try again. The computer was unusable so I had to reinstall everything.

It was on a Fedora 34 trying to install for `5.14`. `uninstall-help` could not be run because there was no network connection so the `dnf` commands would not run. My suspicion is that among the compilation options such as choosing between `generic cpu` and `amd` one, it might have created broken binaries that could not be loaded.

The only problem is that the script does not provide good backup solutions to recover from when most of the drivers are down. I still do not understand why would it affect all of the other kernels as well. are the `/lib/modules` common for all of them? I can also recall that during the local `dnf install` step, it removed a `kernel` package, probably `kernel-modules` and/or `kernel-headers` replacing it with the locally compiled one. Is it possible to create a local backup of that before doing the install so we can install from a local offline backup?

**Comment by AdelKS** [comment]
If it's Fedora then yes, I could not figure out how to have several kernels installed at the same time... 

```
 -> Building successfully finished!
Do you want to install the new Kernel ? y/[n]: y
[sudo] password for adel: 
Last metadata expiration check: 0:16:21 ago on Sun 05 Sep 2021 04:02:02 PM CEST.
Dependencies resolved.
================================================================================
 Package             Arch        Version                Repository         Size
================================================================================
Installing:
 kernel              x86_64      5.14.1_tkg_cfs-1       @commandline       26 M
 kernel-devel        x86_64      5.14.1_tkg_cfs-1       @commandline      135 M
Upgrading:
 kernel-headers      x86_64      5.14.1_tkg_cfs-1       @commandline      1.3 M
Removing:
 kernel              x86_64      5.9.13-200.fc33        @updates            0  
```
So if you install a broken kernel, you will need to repair the system through a live USB + chroot where you reinstall the original kernel with dnf: you do not have to format and reinstall everything. 

We need to at least add a warning about it for the time being. 

If you want to have several kernels installed in Fedora, it is only for "advanced" users: you need use `Generic` as a distro for now, not `Fedora`, in the install script. DKMS modules won't work and the installed kernel will not be tracked by `dnf` but you will be able to chose it separately at boot from the grub menu. You also need to have `dracut` installed.

**Comment by AdelKS** [comment]
Please give a try to the Generic install in the PR #304 and see if it works for you, it should install a kernel on the side with your current one and cannot break your install since if the kernel doesn't start, just reboot and chose the original one on GRUB. The installed files are just not tracked by `dnf`, DKMS should also work.

**Comment by LecrisUT** [comment]
So far your instructions work for not breaking the current kernels. The `anbox` modules still do not work, but now I can play around with the installer

**Comment by AdelKS** [comment]
> So far your instructions work for not breaking the current kernels

Awesome!

> The anbox modules still do not work

Have you had a read at the [README](https://github.com/Frogging-Family/linux-tkg#anbox-usage) on how to use it ? I have never used it myself so I cannot be of further help.

**Comment by LecrisUT** [comment]
Yes, I've read that part of the README. So far, neither `modprobe` nor `ls /dev/{ashmem,binder}` yield anything. So it can be anything at this point. I am trying on a more generically compiled kernel and try to build up from there. I wonder if it's any problem that I'm on x86?

**Comment by AdelKS** [comment]
>  So far, neither modprobe nor ls /dev/{ashmem,binder} yield anything

They wouldn't anyway since in the README it's stated that you have to _manually_ create the `/dev/binderfs` folder (and not `/dev/binder`) and mount a binder kernel sysfs thingy :thinking: 

Do these commands finish successfully ?
```shell
sudo mkdir /dev/binderfs
sudo mount -t binder binder /dev/binderfs
```

**Comment by AdelKS** [comment]
You still need to install `anbox` on top: https://github.com/anbox/anbox/blob/master/docs/install.md where they talk about how to deal with `binderfs` mounts instead.

> In Ubuntu 19.10 the binder driver doesn't create /dev/binder when loaded. That is intentional. Instead it provides support with binderfs (see https://brauner.github.io/2019/01/09/android-binderfs.html) which is instead since PR anbox/anbox#1309

Looks like a good lead.

**Comment by LecrisUT** [comment]
But also ashmem is lacking so I don't think I'm out of the woods yet. Unrelated, but there is a bad command somewhere in the script that builds or git pulls as sudo making it impossible to delete automatically.

**Comment by AdelKS** [comment]
> But also ashmem is lacking so I don't think I'm out of the woods yet.

My understanding is that both ashmem and binder got merged into `binderfs` that offers both functionalities

> Unrelated, but there is a bad command somewhere in the script that builds or git pulls as sudo making it impossible to delete automatically.

Thanks for the feedback, can you give me the error message in the PR discussion if you encounter it again ? I will try to reproduce (I have my ideas on why, though)

**Comment by LecrisUT** [comment]
Progresses: Using all generic settings when building I managed to load ashmem as well as binderfs. I will confirm again with the previous build to see if it's indeed a compilation issue that blocked it or if it was all blocked by mkdir. This should also give the same permission error as before and I'll post it on the PR once I get something.

There are other issues such as the nvidia modules are not ported over, but one step at a time. First I want to confirm that anbox can run like this.

**Comment by AdelKS** [comment]
> There are other issues such as the nvidia modules are not ported over

They need manual intervention by running dkms. I can actually offer in the script to do it for the user as a convenience: I just don't know what is the command to run, I can look that up though.

**Comment by LecrisUT** [comment]
Best option would be to modprobe/dkms all of the available modules on the current kernel and port them over. Yeah, should be a fairly straight-forward command

**Comment by AdelKS** [comment]
Okay I have found [our lord and savior](https://linuxconfig.org/how-to-modify-an-rpm-package-using-rpmrebuild) to change the package name and its conflicts-with fields, _after_ the rpms are created. I will try to implement it and see how it goes :D

**Comment by AdelKS** [comment]
I finished implementing the change that uses `rpmrebuild` in PR #304, please try out that PR and use `Fedora` as distro. Please confirm that nvidia akmod drivers work too. Thanks!

**Comment by LecrisUT** [comment]
I will test it out line-by-line to also figure out which part/package actually broke the installation in the first attempt. I'll make a bit of a note here to remind myself as well what I need to check:
- [ ] Find exactly which kernel package breaks the other kernels `kernel`, `kernel-modules` etc.. Remember to locally backup the pkgs that are being overwritten.
- [ ] Test if `rpmrebuild` fixes the conflict, but also if any files are being overwritten.
- [ ] [CentOS guide](https://wiki.centos.org/HowTos/Custom_Kernel#Installing_the_new_kernel) suggests installing with `dnf localinstall` or `rpm -ivh`.
- [ ] [Fedoraproject guide](https://fedoraproject.org/wiki/Building_a_custom_kernel#Building_a_Kernel_from_the_Fedora_source_tree) suggests uncommenting `# define buildid .local `. Might be only for `fedpkg`.
- [ ] Probably `fedpkg` and other fedora specific pkgs can be ignored.
- [ ] Check that `akmods` can be run successfully on any setup.
- [ ] Check with `modprobe` that all modules are being picked up.
- [ ] For `Generic` build, `sudo grub-mkconfig -o /boot/grub/grub.cfg` should give error in fedora.

Unrelated with current but still on list:
- [ ] Test what causes `read` in [this line](https://github.com/Frogging-Family/linux-tkg/pull/304/files#diff-043df5bdbf6639d7a77e1d44c5226fd7371e5259a1e4df3a0dd5d64c30dca44fR484) to not register properly. Potentially invisible characters and/or leftover input during the previous building step.

**Comment by AdelKS** [comment]
> I will test it out line-by-line to also figure out which part/package actually broke the installation in the first attempt

Remember that in the new PR, old kernels are not replaced and you will be able to reboot into them, as long as grub menu is updated.

> Test if rpmrebuild fixes the conflict, but also if any files are being overwritten.

This is all good to go, the kernel is not replaced. I tested that. And no file will be overwritten, except from the package `kernel-headers` but that is normal and shouldn't matter: the newest versions are back-compatible with older ones and it doesn't depend on the kernel type.

> dnf localinstall

The `localinstall` command [has been deprecated](https://dnf.readthedocs.io/en/latest/command_ref.html#install-command).

> Fedoraproject guide suggests uncommenting # define buildid .local . Might be only for fedpkg.

The entire process of creating kernel RPMs is done by the kernel's script and we will not touch that, except if you want to pass custom parameters. I don't know much about it

> Check that akmods can be run successfully on any setup.

Yes please check that, well, just on one setup is enough. We can wait for feedback.

> Check with modprobe that all modules are being picked up.

I think you are missing something here: when you chose `Fedora` as a distro, the script generates RPM packages, and module rebuilding should be handled automatically like with any kernel package you'd install from the official repos. Although in Fedora 34 they cut down the kernel packages into smaller chunks `kernel-core` and `kernel-modules` but that shouldn't change anything.

> For Generic build, sudo grub-mkconfig -o /boot/grub/grub.cfg should give error in fedora.

`Generic` is not meant for `Fedora`, as there is the `Fedora` option in the script. But yea, I  may need to change that, but it's not very important.

> Test what causes read in this line to not register properly. Potentially invisible characters and/or leftover input during the previous building step.

About that, I need to keep it somewhere in a list, I think I need to "flush" the pipe before every prompt and that should do it.

**Comment by LecrisUT** [comment]
So I couldn't replicate the initial error. I remember that something other than the `kernel` was being deleted, but I can't remember what. It seems that there is no issue with the initial rpm installation method, maybe just add a warning at the installation part so that people make sure only the `kernel` package is overwritten. I would prefer that `linux-tkg` would be tracked together with the system kernels so that we do not overpopulate the `/boot` partition and we do not miss any package dependencies.

Let's close this issue for now, but if someone else manages to reproduce this breakage, we can investigate further.

**Comment by AdelKS** [comment]
It is also bad the the `kernel` package gets replaced by `linux-tkg` because if it doesn't work for whatever reason, the user cannot easily get back to the previous one, if the system doesn't boot to desktop.

The PR #304  I opened should in principle start installing the kernel RPM package alongside the currently installed ones, and not replace them anymore. Please try it and report back to me. This change is a welcome fail safe mechanism.

Please try PR #304, using `Fedora` as distro, and report back if everything works as planned, including `akmods`. This uses the normal approach where RPM packages are generated, and tracked down by `dnf`.

Thanks

**Comment by LecrisUT** [comment]
> It is also bad the the `kernel` package gets replaced by `linux-tkg`

It will only override if it's beyond the limit in `/etc/dnf/dnf.conf` (`installonly_limit`). I have even tested it while overriding including with `kernel-module` override. I am not sure what exactly triggered  the initial errors as I cannot reproduce it no matter how hard I try.

Otherwise #304 works as intended, but I don't have something that would hook to the kernel repo to test it. I thought `akmods` would do that, but either it looks for a different repo, e.g. `kmod-nvidia-5.14.1.tkg-pds` (from my repolist these seem to be autogenerated though), or that's a red hearing and in fact it is some other build step that fails (the log seems to step into `Building for target x86_64`)

At least add the `tkg` kernels to `/etc/dnf/dnf.conf`: `installonlypkgs`. There should be an appropriate command for that. The other repo that install kernels is [`elrepo`](http://elrepo.org/tiki/kernel-ml), maybe there are some other standards used there that should be considered. I haven't had any problems with incompatibility using elrepo kernel.

**Comment by AdelKS** [comment]
> It will only override if it's beyond the limit

Okay I believe I heard something about it then forgot. I need to test that.

So wait, regarding your issue here, how come you could not boot in the official original kernel in your case ? I implicitly assumed it was because your kernel got replaced and you had no easy way to get back the official fedora one for example. I mean if it was installed in parallel, you could have rebooted and chose another kernel, and you'd be good to go.

**Comment by LecrisUT** [comment]
> how come you could not boot in the official original kernel in your case

That is exactly what confused me the most, and I really wished I had documented the output.

The only clue I have is that when kernels are overwritten it's an oldest out first situation, except for the rescue kernel of course. The only possibility I could think of is if the latter kernels developed a dependency on additional kernel modules like `kernel-core` etc. that are removed together with kernel. However I did try that, but still couldn't reproduce.

The other clue is that various drivers didn't work and at the very least those should be available in `kernel` package. I don't know if there's anything that can affect that. Maybe even a grub setting. That installation had artifacts of other nvidia driver installation attempts, but I can't think of what could impact it like that.
